Deprecate DH harder

mattcaswell opened the pull request:
This is very much an early sight WIP PR. It depends on #13136 and #13135 - those commits are included here.

Still to be done:
- There's still more of DH to deprecate, although I think I've solved most problem areas
- make update is doing some strange stuff. Needs investigation
- I had to deprecate some libssl functions/macros because they pass around a DH object. I need to implement EVP friendly replacements
- Better tests of dhparam

levitte commented on ssl/statem/statem_clnt.c:
```diff
@@ -3082,44 +3063,48 @@ static int tls_construct_cke_dhe(SSL *s, WPACKET *pkt)
-    prime_len = BN_num_bytes(DH_get0_p(dh_clnt));
+
+    /* Generate encoding of server key */
+    encoded_pub_len = EVP_PKEY_get1_tls_encodedpoint(ckey, &encoded_pub);
```
I assume you're going to change this call to the new, more generic name?

mattcaswell replied:
Yeah - whichever PR gets there first, I'll update the other one.

t8m commented:
Do we actually need non-deprecated replacement for SSL_set_tmp_dh now that the libssl should automatically choose right safe prime from the known good ones? I propose to not add such replacement.

mattcaswell commented:
> Do we actually need non-deprecated replacement for SSL_set_tmp_dh now that the libssl should automatically choose right safe prime from the known good ones? I propose to not add such replacement.

I believe there are many people who want to use their specific DH params, and would be broken by removing that capability.

t8m commented:
The reason they used specific DH params was mostly that the OpenSSL itself historically did not provide reasonable ones built in. But OK I suppose OpenSSL should still allow users to shoot themselves into their feet.

slontis commented on include/openssl/dh.h:
```diff
@@ -147,7 +147,7 @@ DEPRECATEDIN_3_0(const DH_METHOD *DH_get_default_method(void))
 DEPRECATEDIN_3_0(DH *DH_new_method(ENGINE *engine))
 
-DH *DH_new(void);
+DEPRECATEDIN_3_0(DH *DH_new(void))
```
I don't understand the logic of just deprecating the new..
Why doesnt it also do the free, upref and bits?

mattcaswell replied:
Because this is WIP and as I said in the introduction:

"There's still more of DH to deprecate, although I think I've solved most problem areas"

slontis reviewed (commented):
Do .pod files need updating also?

mattcaswell commented:
> Do .pod files need updating also?

Yes.

slontis commented:
No problems.. I am mainly asking because I am doing the EC deprecation currently - just making sure that was not deliberate.

mattcaswell commented:
I've rebased this to pick up #13136 and #13135 which have now been merged, so this PR is no longer dependent on them.

There's also been numerous updates, including deprecating the remaining DH functions and a number of fixes that resulted. Fixed the "make update" strangeness.

Still to do: documentation updates; more testing

kroeckx commented:
So I'm confused what you think "no-dh" means. To me it means we have no support for DH at all, and that the EVP functions return an error in case you try to use them with DH.

mattcaswell commented:
> So I'm confused what you think "no-dh" means. 

It means there is no low-level DH implementation. It so happens that if you have  no DH implementation then the EVP functions will return an error. Unless you happen to plug in a third party provider that *does* have DH support, in which case it will start to work.

kroeckx commented:
So that means that the SSL library always has DHE support, but
just fails to negiotate a cipher using it? Or does it detect that
DH is not provided, and then not announce it supports it/not
select that as cipher?

kroeckx commented:
This branch PR no longer support reading parameters from stdin, master still support this.

levitte commented on include/openssl/dh.h:
```diff
@@ -185,21 +185,24 @@ DEPRECATEDIN_3_0(int DH_compute_key(unsigned char *key, const BIGNUM *pub_key,
+DEPRECATEDIN_3_0(DH *d2i_DHparams(DH **a, const unsigned char **in, long len))
+DEPRECATEDIN_3_0(int i2d_DHparams(const DH *a, unsigned char **out))
+DEPRECATEDIN_3_0(DH *d2i_DHxparams(DH **a, const unsigned char **in, long len))
+DEPRECATEDIN_3_0(int i2d_DHxparams(const DH *a, unsigned char **out))
```
Why not use `DECLARE_ASN1_ENCODE_FUNCTIONS_only_attr`?

mattcaswell replied:
This was a carry over from an early iteration of this PR where I hadn't quite figured out how I need to do this properly. Now fixed.

t8m commented:
> So that means that the SSL library always has DHE support, but just fails to negiotate a cipher using it? Or does it detect that DH is not provided, and then not announce it supports it/not select that as cipher?

It should be the latter - it should properly detect whether DH is provided or not and not announce/negotiate it if it isn't. Anything else would be a bug.

mattcaswell commented:
> So that means that the SSL library always has DHE support, but
just fails to negiotate a cipher using it? Or does it detect that
DH is not provided, and then not announce it supports it/not
select that as cipher?

Most algorithms (e.g. symmetric ciphers and hashes), it detects up front and does not offer ciphersuites based on that if there is no support. It also detects key exchange algorithms up front and does not offer groups in the supported_groups extension that it cannot support (including DH groups). However, checking the code, it looks to me like it doesn't actually suppress DH ciphersuites if it doesn't have DH support. It should do. I'll raise an issue for that.

mattcaswell commented:
#13184

mattcaswell commented:
> This PR no longer support reading parameters from stdin, master still support this.

Hmm. I just tried this on master and got this output:

````
$ cat dh.pem | openssl dhparam -noout -text
Error, unable to load DH parameters
80C24B4C7A7F0000:error::STORE routines:ossl_store_get0_loader_int:unregistered scheme:crypto/store/store_register.c:240:scheme=file
80C24B4C7A7F0000:error::DECODER routines:decoder_process:BIO lib:crypto/encode_decode/decoder_lib.c:511:
````

Trying the same thing on 1.1.1 works fine. So it looks to me like this is broken even in master.

kroeckx commented:
The difference between this PR and master, is that on master it
still waits for the data while this PR just directly returns an
error.

mattcaswell commented:
> The difference between this PR and master, is that on master it
still waits for the data while this PR just directly returns an
error.

It still seems to fail at the same line in the OSSL_DECODER code though. I've raised an issue (#13185).

levitte commented on include/openssl/dh.h:
```diff
@@ -136,9 +136,9 @@ DECLARE_ASN1_ITEM(DHparams)
+DECLARE_ASN1_DUP_FUNCTION_name_attr(OSSL_DEPRECATEDIN_3_0, DH, DHparams)
+#  endif
 
 DEPRECATEDIN_3_0(const DH_METHOD *DH_OpenSSL(void))
```
I'd like to suggest that you do like I did in #13096, and modify all the deprecations in this file to follow https://github.com/openssl/openssl/blob/master/doc/internal/man7/deprecation.pod ?

mattcaswell replied:
Done.

levitte commented on apps/dhparam.c:
```diff
@@ -191,221 +176,206 @@ int dhparam_main(int argc, char **argv)
+            BIO_printf(bio_err, "Error, unable to load parameters\n");
+            goto end;
+        }
+        OSSL_DECODER_CTX_free(decoderctx);
```
In many other commands, we have dropped support for `-informat` (basically ignoring it for anything other than "engine", where that's interesting), and use `load_key()` and `load_pubkey()`.  So an alternative would be to implement `load_keyparams()` and use that instead.  All you'd have to do after calling that is a check that `EVP_PKEY_is_a(tmppkey, "DH")` returns true.

levitte replied:
(I have to remember to document that when I work on the decoders)

levitte replied:
For an example, see #13191

mattcaswell replied:
Unfortunately its not that simple. I need different processing for DER format than I do PEM format. The reason is that I cannot leave the decoder to auto-detect what is in the file for DER. If I do it sometimes gets it wrong and thinks that a DHX file is DSA file. Therefore I have to explicitly tell it to try and load DH, and then if that fails do it for DHX....but only for DER. Its a bit horrible.

mattcaswell commented:
I more or less have this PR in a reviewable condition. However it has grown to be a bit of monster, so I'm planning to break it up into smaller PRs. There are a few things in here which probably deserve to have separate attention anyway.

mattcaswell commented:
I've now split out various parts of this PR into other smaller PRs. See:

#13227, #13228, #13231, #13233

What remains is mostly just the deprecation itself plus miscellaneous fixups. This PR now depends on the PRs listed above. Once they're in I can rebase this and take it out of WIP.

levitte commented on apps/dhparam.c:
```diff
@@ -191,221 +176,237 @@ int dhparam_main(int argc, char **argv)
+                * the contents of the DER file are. The decoders just try
+                * and guess. Unfortunately with DHX key types they may guess
+                * wrong and think we have a DSA keytype. Therefore we try
+                * both DH and DHX sequentially.
```
Hmmm...  OSSL_STORE in 1.1.1 has functionality that checks for such ambiguity, i.e. it tries all possible decodings and keeps track of how many were successful, and returns an ambiguity error when they are more than one.  That check was never transported to OSSL_DECODER...  it rather just takes the first success and returns that, with no further decoding attempts.

levitte commented:
I edited away the stuff this was pending on that is now merged from the title.

levitte commented:
Please move all the DH specific function names from `doc/man3/d2i_DHparams.pod` and `doc/man3/d2i_X509.pod` to `d2i_RSAPrivateKey.pod`.  And yes, I do realise this means dropping `doc/man3/d2i_DHparams.pod` completely.

mattcaswell commented:
I've rebased this now that all of its dependencies have gone in, and taken it out of WIP.

Please take a look.

slontis reviewed (approved):
Approved subject to travis passing :)

kroeckx commented:
This fixes #13120?

slontis commented:
I assume you are asking that it should have a Fixes #13120 in the commit..

kroeckx commented:
> I assume you are asking that it should have a Fixes #13120 in the commit..

Yes.

levitte commented on include/openssl/evp.h:
```diff
@@ -1232,9 +1232,9 @@ struct dsa_st *EVP_PKEY_get1_DSA(EVP_PKEY *pkey);
-struct dh_st *EVP_PKEY_get1_DH(EVP_PKEY *pkey);
+DEPRECATEDIN_3_0(int EVP_PKEY_set1_DH(EVP_PKEY *pkey, struct dh_st *key))
+DEPRECATEDIN_3_0(struct dh_st *EVP_PKEY_get0_DH(const EVP_PKEY *pkey))
+DEPRECATEDIN_3_0(struct dh_st *EVP_PKEY_get1_DH(EVP_PKEY *pkey))
```
`OSSL_DEPRECATEDIN_3_0`

mattcaswell replied:
Fixed.

mattcaswell commented:
> In this case, you also have d2i_DHparams.pod to deal with. I'd suggest dropping it and adding the symbols from there in d2i_RSAPrivateKey.pod as well.

Done.

Please take another look. Removing the "approval done" label for now.

levitte commented on doc/man3/d2i_RSAPrivateKey.pod:
```diff
@@ -2,7 +2,7 @@
 =begin comment
 
-Any deprecated keypair function from d2i_X509.pod are collected in this file.
+Any deprecated keypair/params d2i function are collected on this page.
```
and i2d

mattcaswell replied:
Fixed.

mattcaswell commented:
Fixup pushed. Please take another look.

mattcaswell commented:
Pushed. Thanks.
